Add support for ZKSync Chains EIP712 transactions

DonGambas opened the pull request:
- added support for creating, signing and sending eip712 transactions on networks that use the zksync stack.

Commit:
eip712tx

Commit:
fix rlp encoding issues

DonGambas committed:
update paymaster validation

DonGambas committed:
remove external rlp lib from deps

rweald commented on example/lib/account_overview_screen.dart:
```diff
@@ -62,6 +63,43 @@ class AccountOverviewScreenState extends State<AccountOverviewScreen> {
     fetchBalance();
   }
 
+  void testZkSyncTx() async {
```
Do we want to fully commit this to our example app given it's with a custom client config? Or should we just include it in the code level docs, and then add in support later once we've more formally created the network config for it?

rweald commented on lib/src/gsn/utils.dart:
```diff
@@ -61,3 +63,29 @@ Web3Client getEthClient(String apiUrl) {
+  return hexWithPrefix;
+}  
+
+Uint8List hexToUint8List(String hex) {
```
Pretty sure this already exists in web3dart? Think it's under crypto called `hexToBytes`. Unless this is specifically different, probably better to use their existing version.

rweald commented on lib/src/wallet.dart:
```diff
@@ -1,13 +1,16 @@
+
```
Might be nice to revert these pure whitespace changes. That way it avoids making it look like code changed in this file, when in reality nothing changed.

rweald commented on lib/src/zksync/eip712Transaction.dart:
```diff
@@ -0,0 +1,160 @@
+
+
+
+class Eip712DomainSeparator {
```
Is there a reason we can't just use the `EIP712Domain` class that we already us in other parts of the codebase? With the optional fields you should be able to just use the ones you want/need. That's what I've done already for the transfer changes to support multiple tokens on base.

rweald committed:
Make rlp library explicit dependency

This is necessary to compile the top level SDK given the new ZKSync
dependency on rlp library import

rweald committed:
Cleanup imports to not import transitive dependencies

rweald committed:
Introduce ZKSyncChain class abstraction

rweald committed:
Move specific sending of zksync eip712 into chain class

I also explicitly re-exported the dependency on the EIP712DomainData
class which would have helped your autocomplete in exmaple find it and
not had us reimplement a very similar interface.

rweald committed:
Cleanup some flutter linter issues with new zksync code

rweald committed:
Rename variable for clarity in example app

rweald committed:
Remove now unused classes from zksync stuff

rweald committed:
Rename class for more explicit zksync scoping

rweald committed:
Create wrapper class for Lens Network Sepolia chain

rweald committed:
Remove rlp library and go back to web3dart rlp import

rweald committed:
Add some documentation to our new public APIs

rweald committed:
Pull example of using lens txn into code docs

DonGambas committed:
update testnet rpc (#66)

rweald committed:
Remove zksync specific stuff from generic example project (#65)

We were using this for manual click testing during development, but it
should not be part of our main repo

rweald committed:
Fix formatting using dart fmt

rweald reviewed (approved):
Need to swap the custom `hexToUint8List` implementation with the web3dart provided one, then this looks good to go.

DonGambas committed:
replace hextouint8list with hextobytes

DonGambas committed:
remove unused files
